README
NAME
    Crypt::XkcdPassword - see http://xkcd.com/936/

SYNOPSIS
     use 5.010;
     use Crypt::XkcdPassword;
 
     say Crypt::XkcdPassword->make_password;

DESCRIPTION
    Yet another password generator module inspired by <http://xkcd.com/936/>.

  Constructor
    *   `new(%attr)`

        Creates a new generator. A single generator can be used to generate as
        many passwords as you like.

  Attributes
    This is a Moo-based class.

    *   `words`

        A string which will be appended to `Crypt::XkcdPassword::Words::` to
        form a class name which will be used as a source for words. The
        default is "EN", which means the class used as a source for words is
        `Crypt::XkcdPassword::Words::EN`.

        `Crypt::XkcdPassword::Words::EN` is a list of 10,000 common English
        words.

        `Crypt::XkcdPassword::Words::EN::Roget` is a list of about 8500 words.
        The words are less questionable, but as there are fewer of them, pass
        phrases will be chosen from a smaller pool, thus slightly more
        guessable.

        Also supplied is "IT", a list of 20,000 common Italian words.

    *   `rng`

        A coderef for generating a random number. The coderef is called and
        passed a single numeric argument. The coderef is expected to generate
        a random, positive integer, smaller than the argument. The default is:

         sub { int(rand($_[0])) }

        Perl's default random number generator is often though insufficient
        for practical cryptography, so you may wish to use another random
        number generator.

  Methods
    *   `make_password($size, $filter)`

        Returns the password as a string.

        $size is the length of the password in words. It defaults to 4. For
        the English dictionary that provides over 47 bits of entropy; for the
        Italian dictionary (which has twice as many words), about 56 bits of
        entropy.

        $filter is a test against which each word is checked. It can be a sub
        returning true if the word is OK, or a regular expression matching OK
        words. Words which are not OK will be excluded from passwords. The
        default is to allow any words found in the provided dictionary.

        For reference, 47 bits of entropy is roughly equivalent to an eight
        digit random case-sensitive alphanumeric password (i.e. 62^8).

        This can be called as an object method, or (if you have no desire to
        change the defaults for the `rng` and `words` attributes) as a class
        method. That is, the first line of the example below is a shortcut for
        the second line:

         say Crypt::XkcdPassword->make_password($size);
         say Crypt::XkcdPassword->new->make_password($size);

        Note that the passphrases returned may not be ASCII-safe, and may
        sometimes be inappropriate for uttering in polite company. See
        Crypt::XkcdPassword::Examples for ways of using $filter to resolve
        this situation.

    *   `chars`

        No-op, provided for compatibility with Data::SimplePassword.

    *   `provider`

        No-op, provided for compatibility with Data::SimplePassword.

BUGS
    Please report any bugs to
    <http://rt.cpan.org/Dist/Display.html?Queue=Crypt-XkcdPassword>.

SEE ALSO
    Crypt::XkcdPassword::Examples - how to do stuff with this module.

    Data::SimplePassword - I borrowed this module's interface, so it should
    mostly be possible to s/Data::SimplePassword/Crypt::XkcdPassword/.

    Crypt::PW44 - similar to this one, but with a smaller list of words.

    <http://xkcd.com/936/>.

AUTHOR
    Toby Inkster <tobyink@cpan.org>.

COPYRIGHT AND LICENCE
    This software is copyright (c) 2012, 2014 by Toby Inkster.

    This is free software; you can redistribute it and/or modify it under the
    same terms as the Perl 5 programming language system itself.

DISCLAIMER OF WARRANTIES
    THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
    WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
    MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



Changes
Crypt-XkcdPassword
==================

Created:      2012-02-02
Home page:    <https://metacpan.org/release/Crypt-XkcdPassword>
Bug tracker:  <http://rt.cpan.org/Dist/Display.html?Queue=Crypt-XkcdPassword>
Maintainer:   Toby Inkster (TOBYINK) <tobyink@cpan.org>

0.008	2014-09-13

 [ Packaging ]
 - Fix issue tracker metadata so that 'NOBODY' doesn't show up in CREDITS
   file for reporting some bugs.

0.007	2014-09-10

 - Added: Bundle `passphrase` command-line tool.
 - Don't depend on Role::Commons.
 - General tidying.
 - Support for Perl 5.8.

0.006	2014-09-10

 [ Packaging ]
 - Switch to Dist::Inkt.

0.005	2013-05-21

 - Any::Moose -> Moo + Types::Standard; smartmatch -> match::smart;
   Class::Load -> Module::Runtime.

0.004	2012-07-06

 - Added: EN::Roget word list.
 - Switch from Moo to Any::Moose.

0.003	2012-03-06

 [ Documentation ]
 - Crypt::XkcdPassword::Examples

 [ Packaging ]
 - Require Perl 5.10.1 (Unicode in 5.10.0 is too broken).

 [ Other ]
 - Added: Provide a $filter parameter to make_password.
   Fixes RT#74685.
   Fixes RT#74684.
   <https://rt.cpan.org/Public/Bug/Display.html?id=74685>
   <https://rt.cpan.org/Public/Bug/Display.html?id=74684>

0.002	2012-02-03

 [ Bug Fixes ]
 - Forgot to run mo-inline in 0.001, creating an undeclared dependency on
   Mo. Have switched to Moo now anyway, and declared that as a dependency.

0.001	2012-02-02	Initial release


lib/Crypt/XkcdPassword.pm
package Crypt::XkcdPassword;

use 5.010;
use Object::AUTHORITY;
use utf8;

BEGIN {
	$Crypt::XkcdPassword::AUTHORITY = 'cpan:TOBYINK';
	$Crypt::XkcdPassword::VERSION   = '0.002';
}

use Carp qw/carp croak/;
use Class::Load qw/try_load_class/;
use Moo;

has rng => (
	is      => 'rw',
	isa     => sub { ref $_[0] eq 'CODE' },
	default => sub { sub { int(rand($_[0])) } },
	);

has words => (
	is      => 'rw',
	isa     => sub { !ref $_[0] },
	default => sub { 'EN' },
	);

*chars = *provider = sub {};

sub make_password
{
	my ($self, $length) = @_;
	
	$self = $self->new unless ref $self;
	
	$length = 4 unless defined $length;
	$length = 4 unless $length > 0;
	
	my $rng        = $self->rng;
	my $words      = $self->_word_list;
	my $word_count = @$words;
	
	return
		join ' ',
		map { $words->[$_] }
		map { $rng->($word_count) }
		1 .. $length;
}

sub _word_list
{
	my ($self) = @_;
	my $class = sprintf 'Crypt::XkcdPassword::Words::%s', $self->words;
	try_load_class($class)
		or do {
			carp "$class could not be loaded, switching to 'EN'";
			croak "No point switching!" if $self->words eq 'EN';
			$self->words('EN');
			return $self->_word_list;
		};
	
	return $class->words;
}

__PACKAGE__
__END__

=head1 NAME

Crypt::XkcdPassword - see http://xkcd.com/936/

=head1 SYNOPSIS

 use 5.010;
 use Crypt::XkcdPassword;
 
 say Crypt::XkcdPassword->make_password;

=head1 DESCRIPTION

Yet another password generator module inspired by L<http://xkcd.com/936/>.

=head2 Constructor

=over

=item * C<< new(%attr) >>

Creates a new generator. A single generator can be used to generate as
many passwords as you like.

=back

=head2 Attributes

This module doesn't use Moose, but uses its "mini-me", L<Mo>. Like Moose,
Mo provides the concept of attributes, which have a getter/setter method.

=over

=item * C<< words >>

A string which will be appended to C<< Crypt::XkcdPassword::Words:: >>
to form a class name which will be used as a source for words. The default
is "EN", which means the class used as a source for words is
C<< Crypt::XkcdPassword::Words::EN >>.

C<< Crypt::XkcdPassword::Words::EN >> is a list of 10,000 common English
words.

Also supplied is "IT", a list of 20,000 common Italian words.

=item * C<< rng >>

A coderef for generating a random number. The coderef is called and passed
a single numeric argument. The coderef is expected to generate a random,
positive integer, smaller than the argument. The default is:

 sub { int(rand($_[0])) }

Perl's default random number generator is often though insufficient for
practical cryptography, so you may wish to use another random number
generator.

=back

=head2 Methods

=over

=item * C<< make_password($size) >>

Returns the password as a string.

$size is the length of the password in words. It defaults to 4. For the
English dictionary that provides over 47 bits of entropy; for the Italian
dictionary (which has twice as many words), about 56 bits of entropy.

For reference, 47 bits of entropy is roughly equivalent to an eight digit
random case-sensitive alphanumeric password (i.e. 62^8).

This can be called as an object method, or (if you have no desire to
change the defaults for the C<rng> and C<words> attributes) as a class
method. That is, the first line of the example below is a shortcut for
the second line:

 say Crypt::XkcdPassword->make_password($size);
 say Crypt::XkcdPassword->new->make_password($size);

Note that the passphrases returned may not be ASCII-safe, and may sometimes
be inappropriate for uttering in polite company.

=item * C<< chars >>

No-op, provided for compatibility with Data::SimplePassword.

=item * C<< provider >>

No-op, provided for compatibility with Data::SimplePassword.

=back

=head1 BUGS

Please report any bugs to
L<http://rt.cpan.org/Dist/Display.html?Queue=Crypt-XkcdPassword>.

=head1 SEE ALSO

L<Data::SimplePassword> - I borrowed this module's interface, so it
should mostly be possible to s/Data::SimplePassword/Crypt::XkcdPassword/.

L<Crypt::PW44> - similar to this one, but with a smaller list of words.

L<http://xkcd.com/936/>.

=head1 AUTHOR

Toby Inkster E<lt>tobyink@cpan.orgE<gt>.

=head1 COPYRIGHT AND LICENCE

This software is copyright (c) 2012 by Toby Inkster.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=head1 DISCLAIMER OF WARRANTIES

THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



Makefile.PL
use strict;
use ExtUtils::MakeMaker 6.17;

my $EUMM = eval( $ExtUtils::MakeMaker::VERSION );

my $meta = {
  "abstract"           => "see http://xkcd.com/936/",
  "author"             => ["Toby Inkster (TOBYINK) <tobyink\@cpan.org>"],
  "dynamic_config"     => 0,
  "generated_by"       => "Dist::Inkt::Profile::TOBYINK version 0.016, CPAN::Meta::Converter version 2.140640",
  "keywords"           => [],
  "license"            => ["perl_5"],
  "meta-spec"          => {
                            url => "http://search.cpan.org/perldoc?CPAN::Meta::Spec",
                            version => 2,
                          },
  "name"               => "Crypt-XkcdPassword",
  "no_index"           => { directory => ["eg", "examples", "inc", "t", "xt"] },
  "prereqs"            => {
                            configure => { requires => { "ExtUtils::MakeMaker" => 6.17 } },
                            runtime => {
                              requires => {
                                "match::simple" => 0,
                                "Module::Runtime" => 0,
                                "Moo" => 0,
                                "perl" => 5.008,
                                "Types::Standard" => 0,
                              },
                            },
                            test => { requires => { "Test::More" => 0.96 } },
                          },
  "provides"           => {
                            "Crypt::XkcdPassword"                   => { file => "lib/Crypt/XkcdPassword.pm", version => 0.008 },
                            "Crypt::XkcdPassword::Words::EN"        => { file => "lib/Crypt/XkcdPassword/Words/EN.pm", version => 0.008 },
                            "Crypt::XkcdPassword::Words::EN::Roget" => {
                                                                         file => "lib/Crypt/XkcdPassword/Words/EN/Roget.pm",
                                                                         version => 0.008,
                                                                       },
                            "Crypt::XkcdPassword::Words::IT"        => { file => "lib/Crypt/XkcdPassword/Words/IT.pm", version => 0.008 },
                          },
  "release_status"     => "stable",
  "resources"          => {
                            bugtracker   => {
                                              web => "http://rt.cpan.org/Dist/Display.html?Queue=Crypt-XkcdPassword",
                                            },
                            homepage     => "https://metacpan.org/release/Crypt-XkcdPassword",
                            license      => ["http://dev.perl.org/licenses/"],
                            repository   => {
                                              type => "git",
                                              url  => "git://github.com/tobyink/p5-crypt-xkcdpassword.git",
                                              web  => "https://github.com/tobyink/p5-crypt-xkcdpassword",
                                            },
                            X_identifier => "http://purl.org/NET/cpan-uri/dist/Crypt-XkcdPassword/project",
                          },
  "version"            => 0.008,
  "x_provides_scripts" => { passphrase => { file => "bin/passphrase" } },
};

my %dynamic_config;

my %WriteMakefileArgs = (
	ABSTRACT   => $meta->{abstract},
	AUTHOR     => ($EUMM >= 6.5702 ? $meta->{author} : $meta->{author}[0]),
	DISTNAME   => $meta->{name},
	VERSION    => $meta->{version},
	EXE_FILES  => [ map $_->{file}, values %{ $meta->{x_provides_scripts} || {} } ],
	NAME       => do { my $n = $meta->{name}; $n =~ s/-/::/g; $n },
	test       => { TESTS => "t/*.t" },
	%dynamic_config,
);

$WriteMakefileArgs{LICENSE} = $meta->{license}[0] if $EUMM >= 6.3001;

sub deps
{
	my %r;
	for my $stage (@_)
	{
		for my $dep (keys %{$meta->{prereqs}{$stage}{requires}})
		{
			next if $dep eq 'perl';
			my $ver = $meta->{prereqs}{$stage}{requires}{$dep};
			$r{$dep} = $ver if !exists($r{$dep}) || $ver >= $r{$dep};
		}
	}
	\%r;
}

my ($build_requires, $configure_requires, $runtime_requires, $test_requires);
if ($EUMM >= 6.6303)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{TEST_REQUIRES}      ||= deps('test');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');
}
elsif ($EUMM >= 6.5503)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build', 'test');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');	
}
elsif ($EUMM >= 6.52)
{
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime', 'build', 'test');	
}
else
{
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('configure', 'build', 'test', 'runtime');	
}

{
	my ($minperl) = reverse sort(
		grep defined && /^[0-9]+(\.[0-9]+)?$/,
		map $meta->{prereqs}{$_}{requires}{perl},
		qw( configure build runtime )
	);
	
	if (defined($minperl))
	{
		die "Installing $meta->{name} requires Perl >= $minperl"
			unless $] >= $minperl;
		
		$WriteMakefileArgs{MIN_PERL_VERSION} ||= $minperl
			if $EUMM >= 6.48;
	}
}

sub FixMakefile
{
	return unless -d 'inc';
	my $file = shift;
	
	local *MAKEFILE;
	open MAKEFILE, "< $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	my $makefile = do { local $/; <MAKEFILE> };
	close MAKEFILE or die $!;
	
	$makefile =~ s/\b(test_harness\(\$\(TEST_VERBOSE\), )/$1'inc', /;
	$makefile =~ s/( -I\$\(INST_ARCHLIB\))/ -Iinc$1/g;
	$makefile =~ s/( "-I\$\(INST_LIB\)")/ "-Iinc"$1/g;
	$makefile =~ s/^(FULLPERL = .*)/$1 "-Iinc"/m;
	$makefile =~ s/^(PERL = .*)/$1 "-Iinc"/m;
	
	open  MAKEFILE, "> $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	print MAKEFILE $makefile or die $!;
	close MAKEFILE or die $!;
}

my $mm = WriteMakefile(%WriteMakefileArgs);
FixMakefile($mm->{FIRST_MAKEFILE} || 'Makefile');
exit(0);

